src/PlusDataCollection/WinProbe/vtkPlusWinProbeVideoSource.cxx

```diff
@@ -1125,37 +1199,90 @@ int32_t vtkPlusWinProbeVideoSource::GetSpatialCompoundCount()
 }
 
 //----------------------------------------------------------------------------
-
-void vtkPlusWinProbeVideoSource::SetBRFEnabled(bool value)
+PlusStatus vtkPlusWinProbeVideoSource::SetExtraSourceMode(Mode mode)
 {
-  if(Connected)
+  if ((GetBRFEnabled() && mode == Mode::BRF) || (GetMModeEnabled() && mode == Mode::M) || (GetARFIEnabled() && mode == Mode::ARFI))
   {
-    if(m_Mode == Mode::M)
-    {
-      SetMIsEnabled(false);
-    }
-    if(value)
-    {
-      SetHandleBRFInternally(false);
-      SetBFRFImageCaptureMode(2);
-    }
-    else
-    {
-      SetHandleBRFInternally(true);
-      SetBFRFImageCaptureMode(0);
-    }
+    return PLUS_SUCCESS;
   }
+  SetMIsEnabled(mode == Mode::M);
+  SetARFIIsEnabled(mode == Mode::ARFI);
 
-  if(value)
+  if (mode == Mode::BRF)
   {
-    m_Mode = Mode::BRF;
+    SetHandleBRFInternally(false);
+    SetBFRFImageCaptureMode(2);
   }
   else
   {
-    m_Mode = Mode::B;
+    SetHandleBRFInternally(true);
+    SetBFRFImageCaptureMode(0);
+  }
+
+  if(mode == Mode::M)
+  {
+    SetMIsRevolving(m_MRevolvingEnabled);
+    SetMPRF(m_MPRF);
+    SetMAcousticLineIndex(m_MLineIndex);
+    ::SetMWidth(m_MWidth);
+    ::SetMAcousticLineCount(m_MAcousticLineCount);
+  }
+  else if (mode == Mode::ARFI)
+  {
+    m_Mode = Mode::ARFI;
+    unsigned bSize = m_SSDecimation * m_PrimaryFrameSize[1] * m_PrimaryFrameSize[0];
+    unsigned arfiDataSize = 1024 * 16 * 64 * 30;
+    unsigned timeblockSize = (4 / 2) * 64 * 30;
+    m_ExtraFrameSize = { bSize + arfiDataSize + timeblockSize, 1, 1 };
+    this->AdjustBufferSizes();
+    std::vector<int32_t> zeroData(m_ExtraFrameSize[0] * m_ExtraFrameSize[1] * m_ExtraFrameSize[2], 0);
+    // add a fake zero-filled frame immediately, because the first frame seems to get lost somehow
+    for(unsigned i = 0; i < m_ExtraSources.size(); i++)
+    {
+      double timestamp = vtkIGSIOAccurateTimer::GetSystemTime();
+      if(m_ExtraSources[i]->AddItem(&zeroData[0],
+                                    US_IMG_ORIENT_FM,
+                                    m_ExtraFrameSize, VTK_INT,
+                                    1, US_IMG_RF_REAL, 0,
+                                    this->FrameNumber,
+                                    timestamp,
+                                    timestamp,
+                                    &m_CustomFields) != PLUS_SUCCESS)
+      {
+        LOG_WARNING("Error adding fake zeros item to ARFI video source " << m_ExtraSources[i]->GetSourceId());
+      }
+      else
+      {
+        LOG_WARNING("Success adding fake zeros item to ARFI video source ");
+      }
+    }
+    LOG_DEBUG("GetARFIIsRFSampleDataCaptureEnabled: " << GetARFIIsRFSampleDataCaptureEnabled());
   }
+  SetPendingRecreateTables(true);
+  LOG_INFO("Mode changed to: " << ModeToString(mode));
+  m_Mode = mode;
+  return PLUS_SUCCESS;
 }
 
+//----------------------------------------------------------------------------
+void vtkPlusWinProbeVideoSource::SetBRFEnabled(bool value)
```

Comment:
I would suggest that this get changed to something like `BRFOn()`. The others would be `MModeOn()` and `ARFIOn()`.

Reply:
To preserve the setter/getter structure here and to allow the user to return to mode B after setting an extra source mode on, I made each of these call the new `SetExtraSourceMode` with `Mode::B` as an arg if `value` is false. I think this should be revisited if `SetMode` is looked at in the future

Reply:
It will have to be confirmed that you can be in `Mode::B` without a stream of errors due to no new frames coming across the extra source port.

Reply:
This is fine